Terminal.Gui/Views/TextMaskField.cs

```diff
@@ -0,0 +1,535 @@
+		public TextMaskField (ustring mask) : this (mask, ustring.Empty) { }
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="TextMaskField"/> class using <see cref="LayoutStyle.Computed"/> positioning.
+		/// </summary>
+		/// <param name="mask">Mask or Regex Pattern</param>
+		/// <param name="masktype">Mask or Regex</param>
+		public TextMaskField (ustring mask, TextMaskType masktype) : this (mask, ustring.Empty, masktype) { }
+
+		/// <summary>
+		/// 
+		/// </summary>
+		/// <param name="mask">Mask or Regex Pattern<see cref="masktype"/></param>
+		/// <param name="text">Initial Value(TODO)</param>
+		/// <param name="masktype">Mask or Regex</param>
+		public TextMaskField (ustring mask, ustring text, TextMaskType masktype = TextMaskType.Mask) : base ()
+		{
+			this.masktype = masktype;
+			Mask = mask;
+			Text = text;
+			Initialize ();
+		}
+
+		void Initialize ()
+		{
+			this.Width = masktype == TextMaskType.Mask ? text.Count : 20; // default width.
+			this.Height = 1;
+			this.CanFocus = true;
+		}
+
+		/// <summary>
+		/// Input value
+		/// </summary>
+		public new ustring Text {
+			get {
+				return ustring.Make (text);
+			}
+			set {
+				text = value != ustring.Empty ? value.ToRuneList () : null;
+
+				SetupText ();
+
+				SetNeedsDisplay ();
+			}
+		}
+
+		/// <summary>
+		/// Mask
+		/// </summary>
+		public ustring Mask {
+			get {
+				return ustring.Make (mask);
+			}
+			set {
+				mask = value.ToRuneList ();
+
+				CompileMask ();
+
+				SetupText ();
+
+				SetNeedsDisplay ();
+			}
+		}
+
+		void SetupText ()
+		{
+			if (text != null) {
+				// TODO: Implement logic to match non-perfect inputs (mask == text)
+				switch (masktype) {
+				case TextMaskType.Mask:
+					if (Validate (text) == false) {
+						text = new List<Rune> ();
+					}
+					break;
+				case TextMaskType.Regex:
+					if (Validate (text) == false && RegexValidateInput == true) {
+						text = new List<Rune> ();
+					}
+					break;
+				}
+				return;
+			}
+
+			text = new List<Rune> ();
+
+			switch (masktype) {
+			case TextMaskType.Mask:
+				for (int i = 0; i < mask.Count; i++) {
+					if (MaskRunes.ContainsKey (mask [i])) {
+						text.Add (EmptyRune);
+					} else {
+						text.Add (mask [i]);
+					}
+					cursorPosition = mask.FindIndex (e => MaskRunes.ContainsKey (e));
+				}
+				break;
+			case TextMaskType.Regex:
+				break;
+			}
+		}
+
+		/// <summary>
+		///  
+		/// </summary>
+		public Rune EmptyRune { get; set; } = ' ';
+
+		/// <summary>
+		/// Character to show for input.
+		/// </summary>
+		public Rune InputRune { get; set; } = '_';
+
+		/// <summary>
+		/// If set, shows the mask. If this is set to false it shows <see cref="InputRune"/>.
+		/// </summary>
+		public bool ShowMaskOnEmpty { get; set; } = false;
+
+		/// <summary>
+		/// Validates input on MaskType.Regex.
+		/// </summary>
+		public bool RegexValidateInput { get; set; } = true;
+
+		///inheritdoc/>
+		public override void PositionCursor ()
```

Comment:
It is not immediately obvious what you are doing here. If it were me, I'd creates some unit tests that clearly show intent (and ensure intent doesn't break).